File: piper/extensions/chrome/src/engine/statements/controls_forEach.js
import { stringify } from "../util"


export default async function(statement, onError) {
    try {
        this.loopFlow = null
        let pointer = statement.pointer
        let list = await this.executeStatement(stringify(pointer.list))
        let loopBlock = stringify(pointer.loopBlock).trim()
        let variableName = stringify(pointer.variable)

        for (let i = 0; i < list.length; i++) {
            this.VARIABLE_STACK[variableName] = list[i]
            if (loopBlock !== "") await this.execute(loopBlock)
            
            if (this.isTerminated) break
            if (this.loopFlow === "BREAK") {
                this.loopFlow = null
                break
            } else if (this.loopFlow === "CONTINUE") {
                this.loopFlow = null
                continue
            }

            if (this.isFunctionReturn !== null) break
        }
    } catch (e) {
        onError(e)
    }
}

File: piper/android/app/src/main/java/io/gopiper/piper/engine/statements/controls_forEach.kt
package io.gopiper.piper.engine.statements

import android.util.Log
import io.gopiper.piper.engine.Piper
import org.json.JSONObject
import java.lang.Exception

suspend fun Piper.executeControlsForEach(statement: JSONObject, onError: (e: Exception) -> Unit) {
    try {
        Log.d("hello", "execute_controls_forEach")
        loopFlow = null
        val pointer = statement.getJSONObject("pointer")
        val list = executeStatement(pointer["list"].toString()) as? List<Any?> ?: listOf()
        val loopBlock = pointer["loopBlock"].toString().trim()
        val variableName = pointer["variable"].toString()

        for (item in list) {
            VARIABLE_STACK[variableName] = item

            if (loopBlock != "") execute(loopBlock)

            if (isTerminated) break
            if (loopFlow == "BREAK") {
                loopFlow = null
                break
            } else if (loopFlow == "CONTINUE") {
                loopFlow = null
                continue
            }

            if (isFunctionReturn != null) break
        }
    } catch (e: Exception) {
        onError(e)
    }
}